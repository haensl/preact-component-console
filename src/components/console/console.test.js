/* eslint-env node, jest */
import { h } from 'preact';
import render from 'preact-render-to-string';
import { shallow, deep } from 'preact-render-spy';

import Line from '../line';
import Console, { DEFAULTS } from './';

jest.useFakeTimers();

const componentToString = (component) =>
  component.toString().split('\n').slice(2).join('\n');

const MAX_CHAR_TIMEOUT = DEFAULTS.console.typing.char.avgMs
  + (DEFAULTS.console.typing.char.deviation * DEFAULTS.console.typing.char.avgMs);

const MAX_NEWLINE_TIMEOUT = DEFAULTS.console.typing.line.delay.avgMs
  + (DEFAULTS.console.typing.line.delay.deviation * DEFAULTS.console.typing.line.delay.avgMs);

describe('Console', () => {
  let tree;
  let component;

  describe('empty', () => {
    beforeEach(() => {
      component = deep(<Console />);
      component.rerender();
      tree = componentToString(component);
    });

    it('renders as expected', () => {
      expect(tree).toMatchSnapshot();
    });

    it('adds the default class to the element', () => {
      expect(tree).toMatch(new RegExp(`class="${DEFAULTS.console.classes.element}"`));
    });

    it('adds the cursor line', () => {
      expect(component.find('Line').length).toBe(1);
    });

    it('adds the ps1', () => {
      expect(component.find('PS1').length).toBe(1);
    });

    it('adds the cursor', () => {
      expect(component.find('Cursor').length).toBe(1);
    });
  });

  describe('writing', () => {
    describe('string', ()  => {
      let line;

      beforeEach(() => {
        line = 'test line.';
        component = deep(<Console lines={ `${line}` } />);
        component.rerender();
        tree = componentToString(component);
      });

      it('renders as expected', () => {
        expect(tree).toMatchSnapshot();
      });

      describe('cursor', () => {
        it('writes the first character', () => {
          expect(component.find('Cursor').attr('char')).toEqual('t');
        });
      });

      describe('line', () => {
        it('writes the first character', () => {
          expect(component.find('Line').attr('content')).toEqual('t');
        });
      });

      describe('after some time', () => {
        beforeEach(() => {
          jest.runTimersToTime(MAX_CHAR_TIMEOUT);
          component.rerender();
          tree = componentToString(component);
        });

        it('renders as expected', () => {
          expect(tree).toMatchSnapshot();
        });

        describe('cursor', () => {
          it('writes the second character', () => {
            expect(component.find('Cursor').attr('char')).toEqual('e');
          });
        });

        describe('line', () => {
          it('writes the first two characters', () => {
            expect(component.find('Line').attr('content')).toEqual('te');
          });
        });

        describe('after the whole line has been written', () => {
          beforeEach(() => {
            for(let i = 2; i < line.length; i++) {
              jest.runTimersToTime(MAX_CHAR_TIMEOUT);
              component.rerender();
            }

            tree = componentToString(component);
          });

          it('renders as expected', () => {
            expect(tree).toMatchSnapshot();
          });

          describe('cursor', () => {
            it('writes the last character', () => {
              expect(component.find('Cursor').attr('char')).toEqual('.');
            });
          });

          describe('line', () => {
            it('contains the whole line', () => {
              expect(component.find('Line').attr('content')).toEqual(line);
            });
          });
        });
      });
    });

    describe('array', () => {
      let lines;
      let props;
      let component;
      let line;
      let currentLine;

      beforeEach(() => {
        lines  = [
          'test line.',
          'another line'
        ];
        props = {
          lines
        };
        component = deep(<Console { ...props } />);
      });

      describe('first line', () => {
        beforeEach(() => {
          currentLine = component.state().console.currentLine;
          line = lines[currentLine];
          component.rerender();
          tree = componentToString(component);
        });

        it('renders as expected', () => {
          expect(tree).toMatchSnapshot();
        });

        describe('cursor', () => {
          it('writes the first character', () => {
            expect(component.find('Cursor').attr('char')).toEqual(line.slice(0, 1));
          });
        });

        describe('line', () => {
          it('writes the first character', () => {
            expect(component.find('Line').attr('content')).toEqual(line.slice(0, 1));
          });
        });

        describe('after some time', () => {
          beforeEach(() => {
            jest.runTimersToTime(MAX_CHAR_TIMEOUT);
            component.rerender();
            tree = componentToString(component);
          });

          it('renders as expected', () => {
            expect(tree).toMatchSnapshot();
          });

          describe('cursor', () => {
            it('writes the second character', () => {
              expect(component.find('Cursor').attr('char')).toEqual(line.slice(1, 2));
            });
          });

          describe('line', () => {
            it('writes the first two characters', () => {
              expect(component.find('Line').attr('content')).toEqual(line.slice(0, 2));
            });
          });

          describe('after the whole line has been written', () => {
            beforeEach(() => {
              for(let i = 2; i < line.length; i++) {
                jest.runTimersToTime(MAX_CHAR_TIMEOUT);
                component.rerender();
              }

              tree = componentToString(component);
            });

            it('renders as expected', () => {
              expect(tree).toMatchSnapshot();
            });

            describe('cursor', () => {
              it('writes the last character', () => {
                expect(component.find('Cursor').attr('char')).toEqual(line.slice(-1));
              });
            });

            describe('line', () => {
              it('contains the whole line', () => {
                expect(component.find('Line').attr('content')).toEqual(line);
              });
            });

          describe('second line', () => {
            beforeEach(() => {
              jest.clearAllTimers();
              //jest.runTimersToTime(MAX_NEWLINE_TIMEOUT + MAX_CHAR_TIMEOUT + 200);
              //jest.runOnlyPendingTimers();
              currentLine = component.state().console.currentLine;
              line = lines[currentLine];
              console.log(lines);
              console.log(line);
              component.rerender();
              console.log(JSON.stringify(component.state(), null, 2));
              tree = componentToString(component);
            });

            it('increments the current line', () => {
              expect(currentLine).toBe(1);
            })

            it('renders as expected', () => {
              expect(tree).toMatchSnapshot();
            });

            describe('cursor', () => {
              it('writes the first character', () => {
                expect(component.find('Cursor').attr('char')).toEqual(line.slice(0, 1));
              });
            });

            describe('line', () => {
              it('writes the first character', () => {
                expect(component.find('Line').attr('content')).toEqual(line.slice(0, 1));
              });
            });

            describe('after some time', () => {
              beforeEach(() => {
                jest.runTimersToTime(MAX_CHAR_TIMEOUT);
                component.rerender();
                tree = componentToString(component);
              });

              it('renders as expected', () => {
                expect(tree).toMatchSnapshot();
              });

              describe('cursor', () => {
                it('writes the second character', () => {
                  expect(component.find('Cursor').attr('char')).toEqual(line.slice(1, 2));
                });
              });

              describe('line', () => {
                it('writes the first two characters', () => {
                  expect(component.find('Line').attr('content')).toEqual(line.slice(0, 2));
                });
              });

              describe('after the whole line has been written', () => {
                beforeEach(() => {
                  for(let i = 2; i < line.length; i++) {
                    jest.runTimersToTime(MAX_CHAR_TIMEOUT);
                    component.rerender();
                  }

                  tree = componentToString(component);
                });

                it('renders as expected', () => {
                  expect(tree).toMatchSnapshot();
                });

                describe('cursor', () => {
                  it('writes the last character', () => {
                    expect(component.find('Cursor').attr('char')).toEqual(line.slice(-1));
                  });
                });

                describe('line', () => {
                  it('contains the whole line', () => {
                    expect(component.find('Line').attr('content')).toEqual(line);
                  });
                });
              });
            });
          });

          });

        });
      });
    });
  });
});
