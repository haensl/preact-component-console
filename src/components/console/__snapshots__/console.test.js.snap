// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Console empty renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\"></span>
    <span class=\\"console-cursor\\"></span>
  </pre>
</div>
"
`;

exports[`Console writing array first line after some time after the whole line has been written renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">test line.</span>
    <span class=\\"console-cursor console-cursor--write\\">.</span>
  </pre>
</div>
"
`;

exports[`Console writing array first line after some time after the whole line has been written second line after some time after the whole line has been written renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">test line.</span>
    <span class=\\"console-cursor console-cursor--blink\\"></span>
  </pre>
</div>
"
`;

exports[`Console writing array first line after some time after the whole line has been written second line after some time renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">test line.</span>
    <span class=\\"console-cursor console-cursor--write\\">.</span>
  </pre>
</div>
"
`;

exports[`Console writing array first line after some time after the whole line has been written second line renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">test line.</span>
    <span class=\\"console-cursor console-cursor--write\\">.</span>
  </pre>
</div>
"
`;

exports[`Console writing array first line after some time renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">te</span>
    <span class=\\"console-cursor console-cursor--write\\">e</span>
  </pre>
</div>
"
`;

exports[`Console writing array first line renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">t</span>
    <span class=\\"console-cursor console-cursor--write\\">t</span>
  </pre>
</div>
"
`;

exports[`Console writing string after some time after the whole line has been written renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">test line.</span>
    <span class=\\"console-cursor console-cursor--write\\">.</span>
  </pre>
</div>
"
`;

exports[`Console writing string after some time renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">te</span>
    <span class=\\"console-cursor console-cursor--write\\">e</span>
  </pre>
</div>
"
`;

exports[`Console writing string renders as expected 1`] = `
"<div class=\\"console\\">
  <pre class=\\"console-line\\">
    <span class=\\"console-ps1\\">$</span>
    <span class=\\"console-line-content\\">t</span>
    <span class=\\"console-cursor console-cursor--write\\">t</span>
  </pre>
</div>
"
`;
